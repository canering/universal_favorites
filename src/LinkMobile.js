/*global chrome browser*/
import React, { useEffect } from "react";
import { useDrag, useDrop } from "react-dnd";
import { getEmptyImage } from "react-dnd-html5-backend";
import PropTypes from "prop-types";
import { browserName } from "react-device-detect";

function FolderSubDrop(props) {
    const [, drop] = useDrop({
        accept:"LINK",
        drop({ id: draggingId }){
            props.dropElement(props.id, draggingId, true);
        },
    });

    return(
        <div ref={drop} className="menu-item-content">
            <img src="https://img.icons8.com/color/48/000000/folder-invoices.png" alt="folder" className="icon" />
            {props.label}
        </div>
    );
}

FolderSubDrop.propTypes = {
    label:PropTypes.string,
    dropElement:PropTypes.func,
    id:PropTypes.number
};

function LinkMobile(props) {
    const [, drag, preview] = useDrag({
        item: { type: "LINK" },
        begin(){
            props.saveDraggingId(props.id);
            return {type: "LINK", label:props.item.label, link:props.item.link, draggingType:props.item.type, id:props.id};
        },
    });

    const [{isOver}, drop] = useDrop({
        accept:"LINK",
        drop({ id: draggingId }, monitor){
            if(!monitor.didDrop()) {
                props.dropElement(props.id, draggingId, false);
            }
        },
        collect: monitor => ({
            isOver: monitor.isOver({ shallow: true })
        }),
    });

    useEffect(() => {
        preview(getEmptyImage(), { captureDraggingState: true });
    });

    let style = {};
    if(props.item.clicked) {
        style.backgroundColor = "lightgray";
    }
    if(isOver) {
        if(props.id < props.draggingId) {
            style.borderTop ="2px dashed gray";
        }
        else if(props.id > props.draggingId) {
            style.borderBottom = "2px dashed gray";
        }
    }

    let label = props.item.label;
    if(label.length > 32){
        label = label.substring(0, 31);
        label += "...";
    }

    if(props.item.type === "folder"){
        return (
            <div
                ref={node => drop(node)}
                className="menu-item"
                style={style}
                onClick={(event) => clickHandler(event,
                    () => props.openFolder(props.item.data, props.id))}
            >
                <FolderSubDrop label={label} dropElement={props.dropElement} id={props.id} />
                <img src="https://img.icons8.com/metro/26/000000/drag-reorder.png" alt="drag section" ref={node => drag(node)} className="icon dragIcon" />
            </div>);
    }

    let url = props.item.link;
    if(!url.startsWith("http://") && !url.startsWith("https://")) {
        url = "http://" + url;
    }

    return(
        <div 
            ref={node => drop(node)}
            className="menu-item" 
            style={style}
            onClick={(event) => clickHandler(event, () => {
                if(browserName === "Firefox" || browserName === "Edge") {
                    browser.tabs.create({url:url});
                } else {
                    chrome.tabs.create({url:url});
                }
            })}
        >
            <div className="menu-item-content">
                <img src={"https://www.google.com/s2/favicons?domain=" + props.item.link} alt="link" 
                    className="icon" 
                    onError={getFallbackFavicon} />
                {label}
                <img src="https://img.icons8.com/metro/26/000000/drag-reorder.png" alt="drag section" ref={node => drag(node)} className="icon dragIcon" />
            </div>
        </div>);

    function clickHandler(event, callback) {
        if(event.ctrlKey) {
            props.setClicked(props.id);
        } else if(event.shiftKey) {
            props.setShiftClicked(props.id);
        } else {
            callback();
        }
    }
}

LinkMobile.propTypes = {
    item:PropTypes.object,
    openFolder:PropTypes.func,
    id:PropTypes.number,
    saveDraggingId:PropTypes.func,
    dropElement:PropTypes.func,
    draggingId:PropTypes.number,
    setClicked:PropTypes.func,
    setShiftClicked:PropTypes.func
};

function getFallbackFavicon(event) {
    event.target.src = "https://img.icons8.com/material/24/000000/internet.png";
}

export default LinkMobile;