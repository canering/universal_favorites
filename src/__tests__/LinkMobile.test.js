import React from "react";
import {mount, configure} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import LinkMobile from "../LinkMobile";
import chrome from "sinon-chrome";
import { DndProvider } from "react-dnd";
import toJson from "enzyme-to-json";
import TestBackend from "react-dnd-test-backend";

window.chrome = chrome;

configure({adapter: new Adapter()});

const props = {
    openFolder: jest.fn(),
    id: 0,
    saveDraggingId: jest.fn(),
    dropElement: jest.fn(),
    draggingId: 0,
    setClicked: jest.fn(),
    setShiftClicked: jest.fn()
};

const linkItem = {
    type:"link",
    link:"http://example.com",
    label: "example.com"
};

const folderItem = {
    type:"folder",
    label: "example.com"
};

var linkComponent;
var folderComponent;

beforeEach(() => {
    linkComponent = mount(
        <DndProvider backend={TestBackend}>
            <LinkMobile item={linkItem} {...props} />
        </DndProvider>
    );

    folderComponent = mount(
        <DndProvider backend={TestBackend}>
            <LinkMobile item={folderItem} {...props} />
        </DndProvider>
    );
});

test("link snapshot test", () => {

    let tree = toJson(linkComponent);
    expect(tree).toMatchSnapshot();

    const clickedComponent = mount(
        <DndProvider backend={TestBackend}>
            <LinkMobile item={Object.assign({}, {clicked:true}, linkItem)} {...props} />
        </DndProvider>
    );

    tree = toJson(clickedComponent);
    expect(tree).toMatchSnapshot();
});

test("folder snapshot test", () => {
    let tree = toJson(folderComponent);
    expect(tree).toMatchSnapshot();

    const clickedComponent = mount(
        <DndProvider backend={TestBackend}>
            <LinkMobile item={Object.assign({}, {clicked:true}, folderItem)} {...props} />
        </DndProvider>
    );

    tree = toJson(clickedComponent);
    expect(tree).toMatchSnapshot();
});

test("when folder is clicked openFolder should be called", () => {

    folderComponent.find(".menu-item").prop("onClick")({});

    expect(props.openFolder).toHaveBeenCalledTimes(1);
    expect(props.openFolder).toHaveBeenCalledWith(folderItem, props.id);
});