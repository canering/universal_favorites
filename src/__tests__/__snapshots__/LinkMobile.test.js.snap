// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`folder snapshot test 1`] = `
<DndProvider
  backend={[Function]}
>
  <LinkMobile
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "label": "example.com",
        "type": "folder",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      style={Object {}}
    >
      <FolderSubDrop
        dropElement={[MockFunction]}
        id={0}
        label="example.com"
      >
        <div
          className="menu-item-content"
        >
          <img
            alt="folder"
            className="icon"
            src="folder.png"
          />
          example.com
        </div>
      </FolderSubDrop>
      <img
        alt="drag section"
        className="icon dragIcon"
        src="mobile_drag.png"
      />
    </div>
  </LinkMobile>
</DndProvider>
`;

exports[`folder snapshot test 2`] = `
<DndProvider
  backend={[Function]}
>
  <LinkMobile
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "clicked": true,
        "label": "example.com",
        "type": "folder",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <div
      className="menu-item"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "lightgray",
        }
      }
    >
      <FolderSubDrop
        dropElement={[MockFunction]}
        id={0}
        label="example.com"
      >
        <div
          className="menu-item-content"
        >
          <img
            alt="folder"
            className="icon"
            src="folder.png"
          />
          example.com
        </div>
      </FolderSubDrop>
      <img
        alt="drag section"
        className="icon dragIcon"
        src="mobile_drag.png"
      />
    </div>
  </LinkMobile>
</DndProvider>
`;

exports[`link snapshot test 1`] = `
<DndProvider
  backend={[Function]}
>
  <LinkMobile
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "label": "example.com",
        "link": "http://example.com",
        "type": "link",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <a
      href="http://example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        className="menu-item"
        style={Object {}}
      >
        <img
          alt="link"
          className="icon"
          hidden={false}
          src="default.png"
        />
        <img
          alt="link"
          className="icon"
          hidden={true}
          onLoad={[Function]}
          src="https://www.google.com/s2/favicons?domain=http://example.com"
        />
        example.com
        <img
          alt="drag section"
          className="icon dragIcon"
          src="mobile_drag.png"
        />
      </div>
    </a>
  </LinkMobile>
</DndProvider>
`;

exports[`link snapshot test 2`] = `
<DndProvider
  backend={[Function]}
>
  <LinkMobile
    draggingId={0}
    dropElement={[MockFunction]}
    id={0}
    item={
      Object {
        "clicked": true,
        "label": "example.com",
        "link": "http://example.com",
        "type": "link",
      }
    }
    openFolder={[MockFunction]}
    saveDraggingId={[MockFunction]}
    setClicked={[MockFunction]}
    setShiftClicked={[MockFunction]}
  >
    <a
      href="http://example.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      <div
        className="menu-item"
        style={
          Object {
            "backgroundColor": "lightgray",
          }
        }
      >
        <img
          alt="link"
          className="icon"
          hidden={false}
          src="default.png"
        />
        <img
          alt="link"
          className="icon"
          hidden={true}
          onLoad={[Function]}
          src="https://www.google.com/s2/favicons?domain=http://example.com"
        />
        example.com
        <img
          alt="drag section"
          className="icon dragIcon"
          src="mobile_drag.png"
        />
      </div>
    </a>
  </LinkMobile>
</DndProvider>
`;
